### PR TITLE
Move txPool work from Processing loop event to txPool thread

### PR DESCRIPTION
**Pull request by benaadams**
## Changes

- Move clearing caches to txPool processing new head rather than main processing informing txPool of new head

2.9% of archive since since its processing blocks so fast 
image

#### What types of changes does your code introduce?

- [x] Optimization

#### Requires testing

- [x] Yes

#### If yes, did you write tests?

- [x] Yes (changed existing tests)

### CODE REVIEW

File: src/Nethermind/Nethermind.TxPool/TxPool.cs
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.TxPool/TxPool.cs:204]: Having this event here feels super-weird. Can we at least make it somehow work in txpool context?

Comment (benaadams) [reply src/Nethermind/Nethermind.TxPool/TxPool.cs]: Rename to `TxPoolHeadChanged`?

Comment (benaadams) [reply src/Nethermind/Nethermind.TxPool/TxPool.cs]: Also used the event

File: src/Nethermind/Nethermind.TxPool/TxPool.cs
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.TxPool/TxPool.cs:197]: Could that lead to some inconsistencies in TxPool, so it could give us stale data in normal (not archive) processing mode?

Comment (benaadams) [reply src/Nethermind/Nethermind.TxPool/TxPool.cs:197]: Shouldn't do as we do the bucket Update etc on the other half where this moves to

File: src/Nethermind/Nethermind.TxPool/TxPool.cs
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.TxPool/TxPool.cs:196]: Different question, can we make those two use fixed memory?
Basically we should be able to allocate something ~max tx pool count and only update it when needed, instead of allocating it all the time.

Comment (benaadams) [reply src/Nethermind/Nethermind.TxPool/TxPool.cs:196]: Don't think so easily; txPool is quite a complicated set of data structures and don't know when the snapshot has finished being used (to change it)

However could use it in one more place

Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.TxPool/TxPool.cs:196]: Maybe at least pool them?

Comment (benaadams) [reply src/Nethermind/Nethermind.TxPool/TxPool.cs:196]: They don't have a very good lifecycle as are multi-use

Would need to introduce a ref counted ArrayPoolList type thing.

Out of scope for this PR though?

Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.TxPool/TxPool.cs:196]: eh ok :(

Comment (benaadams) [reply src/Nethermind/Nethermind.TxPool/TxPool.cs:196]: Something to look at with a bigger refactor of txpool

Comment (benaadams) [reply src/Nethermind/Nethermind.TxPool/TxPool.cs:196]: Something to look at with a bigger refactor of txpool vs removing time impact on main processing loop

**Review by LukaszRozmej — APPROVED** [review]
Let's wait for @marcindsobczak approval before merging

**Comment by benaadams** [comment]
Aside this also fix JsonRpc tests timing out

**Review by marcindsobczak — APPROVED** [review]
LGTM, worth to run on some node/nodes and observe for a few days

**Pull request merged**
